In [9]:
import numpy as np
import pandas as pd
from pathlib import Path

In [10]:
train_df = pd.read_csv(Path('Resources/2019loans.csv'))
test_df = pd.read_csv(Path('Resources/2020Q1loans.csv'))

In [11]:
train_df

loan_amnt  int_rate  installment home_ownership  annual_inc  \
0         7000.0    0.1894       256.38       MORTGAGE     75000.0   
1        40000.0    0.1614       975.71       MORTGAGE    102000.0   
2        11000.0    0.2055       294.81           RENT     45000.0   
3         4000.0    0.1612       140.87       MORTGAGE     38000.0   
4        14000.0    0.1797       505.93       MORTGAGE     43000.0   
...          ...       ...          ...            ...         ...   
12175    19975.0    0.2565       801.09           RENT     28000.0   
12176    15000.0    0.1774       540.34           RENT     50000.0   
12177     3600.0    0.1862       131.28           RENT     60000.0   
12178    15000.0    0.0881       475.68       MORTGAGE     62000.0   
12179    15000.0    0.1774       540.34           RENT     54080.0   

      verification_status pymnt_plan    dti  delinq_2yrs  inq_last_6mths  ...  \
0            Not Verified          n  28.62          0.0             2.0  ...   
1         Source Verified          n  11.72          2.0             0.0  ...   
2                Verified          n  37.25          1.0             3.0  ...   
3            Not Verified          n  42.89          1.0             0.0  ...   
4         Source Verified          n  22.16          1.0             0.0  ...   
...                   ...        ...    ...          ...             ...  ...   
12175        Not Verified          n  28.42          0.0             0.0  ...   
12176            Verified          n  23.43          4.0             0.0  ...   
12177        Not Verified          n  28.80          0.0             1.0  ...   
12178     Source Verified          n  11.44          0.0             0.0  ...   
12179        Not Verified          n  21.66          1.0             1.0  ...   

       percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0                  87.5                   0.0        0.0         352260.0   
1                   0.0                   0.0        0.0         294664.0   
2                   7.7                   0.0        0.0          92228.0   
3                 100.0                   0.0        0.0         284273.0   
4                  25.0                   0.0        0.0         120280.0   
...                 ...                   ...        ...              ...   
12175              16.7                   0.0        0.0          50055.0   
12176              11.1                   0.0        0.0          70324.0   
12177               0.0                   0.0        0.0          83765.0   
12178               0.0                   0.0        0.0         189930.0   
12179             100.0                   0.0        0.0          48968.0   

      total_bal_ex_mort  total_bc_limit  total_il_high_credit_limit  \
0               62666.0         35000.0                     10000.0   
1              109911.0          9000.0                     71044.0   
2               36007.0         33000.0                     46328.0   
3               52236.0         13500.0                     52017.0   
4               88147.0         33300.0                     78680.0   
...                 ...             ...                         ...   
12175           28192.0         18700.0                     19055.0   
12176           57025.0         13300.0                     54824.0   
12177           55156.0         14800.0                     53065.0   
12178           23748.0          7000.0                     32930.0   
12179           29786.0          2900.0                     33968.0   

       hardship_flag  debt_settlement_flag     target  
0                  N                     N   low_risk  
1                  N                     N   low_risk  
2                  N                     N   low_risk  
3                  N                     N   low_risk  
4                  N                     N   low_risk  
...              ...                   ...        ...  
12175              N   

In [12]:
X_train = pd.get_dummies(train_df.drop(columns=['target']))
y_train = train_df['target']

In [13]:
X_train

loan_amnt  int_rate  installment  annual_inc    dti  delinq_2yrs  \
0         7000.0    0.1894       256.38     75000.0  28.62          0.0   
1        40000.0    0.1614       975.71    102000.0  11.72          2.0   
2        11000.0    0.2055       294.81     45000.0  37.25          1.0   
3         4000.0    0.1612       140.87     38000.0  42.89          1.0   
4        14000.0    0.1797       505.93     43000.0  22.16          1.0   
...          ...       ...          ...         ...    ...          ...   
12175    19975.0    0.2565       801.09     28000.0  28.42          0.0   
12176    15000.0    0.1774       540.34     50000.0  23.43          4.0   
12177     3600.0    0.1862       131.28     60000.0  28.80          0.0   
12178    15000.0    0.0881       475.68     62000.0  11.44          0.0   
12179    15000.0    0.1774       540.34     54080.0  21.66          1.0   

       inq_last_6mths  open_acc  pub_rec  revol_bal  ...  \
0                 2.0      20.0      0.0    40414.0  ...   
1                 0.0      10.0      0.0    43531.0  ...   
2                 3.0      23.0      0.0     8242.0  ...   
3                 0.0       7.0      0.0    12767.0  ...   
4                 0.0      22.0      0.0    11182.0  ...   
...               ...       ...      ...        ...  ...   
12175             0.0      15.0      0.0     9679.0  ...   
12176             0.0      16.0      0.0     3193.0  ...   
12177             1.0      14.0      0.0    11804.0  ...   
12178             0.0       5.0      0.0      157.0  ...   
12179             1.0       5.0      0.0     8349.0  ...   

       verification_status_Verified  pymnt_plan_n  initial_list_status_f  \
0                                 0             1                      1   
1                                 0             1                      0   
2                                 1             1                      0   
3                                 0             1                      0   
4                                 0             1                      0   
...                             ...           ...                    ...   
12175                             0             1                      0   
12176                             1             1                      0   
12177                             0             1                      0   
12178                             0             1                      0   
12179                             0             1                      1   

       initial_list_status_w  application_type_Individual  \
0                          0                            1   
1                          1                            1   
2                          1                            1   
3                          1                            0   
4                          1                            1   
...                      ...                          ...   
12175                      1                            1   
12176                      1                            1   
12177                      1                            1   
12178                      1                            0   
12179                      0                            1   

       application_type_Joint App  hardship_flag_N  hardship_flag_Y  \
0                               0                1                0   
1                               0                1                0   
2                               0                1                0   
3                               1                1                0   
4                               0                1                0   
...                           ...              ...              ...   
12175                           0                1                0   
12176                           0                1                0   
12177                           0                1                0   
12178                        

In [14]:
X_test = pd.get_dummies(test_df.drop(columns=['target']))
y_test = test_df['target']

In [15]:
# add missing dummy variables to testing set
for col in X_train.columns:
    if col not in X_test.columns:
        X_test[col] = 0

In [16]:
# Train the Logistic Regression model on the unscaled data and print the model score
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression(solver='lbfgs', random_state=1)
clf.fit(X_train, y_train)
clf.score(X_test, y_test)

Z:\Travis\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.5076563164610803

In [17]:
# Train a Random Forest Classifier model
# and print the model score
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier()
clf.fit(X_train, y_train)
clf.score(X_test, y_test)

0.6388770735857082

The RandomForestClassifier is the more accurate model with a 64% accuracy score, versus the 51% accuracy score of the LogisticRegression model.

In [18]:
# Scale the data
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [19]:
# Train the Logistic Regression model on the scaled data and print the model score
from sklearn.linear_model import LogisticRegression
clf = LogisticRegression(solver='lbfgs')
clf.fit(X_train_scaled, y_train)
clf.score(X_test_scaled, y_test)

Z:\Travis\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:765: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.7598894087622289

In [20]:
# Train a Random Forest Classifier model on the scaled data and print the model score
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier()
clf.fit(X_train_scaled, y_train)
clf.score(X_test_scaled, y_test)

0.6482347937048064

On the scaled data, the LogisticRegression increases its accuracy to 76%. The RandomForestClassifier remains relatively unchanged with a 65% accuracy score. This shows that the LogisticRegression is sensitive to how the data is scaled, while random forests are not. According to [this paper](https://scholar.smu.edu/cgi/viewcontent.cgi?article=1041), when there are 30 or more explanatory variables, logistic regressions outperform random forest classifiers. The data here has 84 explanatory variables, which aligns with the results of the paper.